tests/e2e/s3/utils.py

```diff
@@ -0,0 +1,219 @@
+    namespace = os.environ.get("NAMESPACE")
+    update_agent_biling_clients(
+        {
+            "cpuMetricName": "effective_compute_seconds",
+            "activeTimeMetricName": "active_time_seconds",
+            "collectEverySeconds": 5,
+            "accumulateEverySeconds": 5,
+            "clients": {
+                "s3": {
+                    "bucket": "metrics",
+                    "region": "us-east-1",
+                    "prefixInBucket": "autoscaler-agent",
+                    "endpoint": "http://minio-service.%s:9000" % namespace,
+                    "pushEverySeconds": 10,
+                    "pushRequestTimeoutSeconds": 20,
+                    "maxBatchSize": 200,
+                }
+            },
+        }
+    )
+
+
+def agent_remove_s3():
+    update_agent_biling_clients(
+        {
+            "cpuMetricName": "effective_compute_seconds",
+            "activeTimeMetricName": "active_time_seconds",
+            "collectEverySeconds": 5,
+            "accumulateEverySeconds": 5,
+            "clients": {},
+        }
+    )
+
+
+def agent_restart():
+    os.system("kubectl rollout restart daemonset autoscaler-agent -n kube-system")
+
+
+def s3_create_bucket(local_endpoint):
+    s3 = boto3.client("s3", endpoint_url=local_endpoint)
+    s3.create_bucket(Bucket="metrics")
+
+
+def s3_check_file(local_endpoint):
+    s3 = boto3.client("s3", endpoint_url=local_endpoint)
+    found = False
+    for i in range(100):
+        response = s3.list_objects_v2(Bucket="metrics")
+        if "Contents" in response:
+            found = True
+            break
+        time.sleep(1)
+        print("Waiting for the file in s3")
+
+    if not found:
+        raise Exception("File not found in s3")
+
+    assert len(response["Contents"]) > 0
+    key = response["Contents"][0]["Key"]
+
+    # Example key:
+    # autoscaler-agent/year=2024/month=04/day=12/07:19:27Z_DGFMM7kfyLFdJASCnJfEYT.ndjson.gz
+    assert key.startswith("autoscaler-agent")
+    assert key.endswith(".ndjson.gz")
+
+    response = s3.get_object(Bucket="metrics", Key=key)
+    result = subprocess.check_output(
+        ["gzip", "-d", "-c"], input=response["Body"].read()
+    )
+    result = json.loads(result)
+    print(result)
+
+    assert "events" in result
+
+    events = result["events"]
+    assert len(events) > 0
+
+    # Example events[0]:
+    #
+    # {
+    #   "idempotency_key": "2024-04-12T12:10:00.845237Z-autoscaler-agent-nwcnd-1/4",
+    #   "metric": "effective_compute_seconds",
+    #   "type": "incremental",
+    #   "endpoint_id": "test",
+    #   "start_time": "2024-04-12T12:09:55.844007537Z",
+    #   "stop_time": "2024-04-12T12:10:00.845237402Z",
+    #   "value": 1
+    # }
+    assert set(events[0].keys()) == set(
+        [
+            "idempotency_key",
+            "metric",
+            "type",
+            "endpoint_id",
+            "start_time",
+            "stop_time",
+            "value",
+        ]
+    ), events[0].keys()
```

Comment:
IMO: we're unlikely to mess up the JSON fields. We _may_ be likely to mess up the values we send - it'd be good to check those (at least, make sure that the start/stop times are self-consistent, and the values make sense for those start/stop times)

Reply:
Hm, I think it's ok enough for now — but a good follow-up would be more exhaustive checking of all of the contents. Basically thinking that we should check that every value we receive makes sense (i.e. timestamps line up right, metric values make sense for the timestamps).

WDYT?

Reply:
I've added a schema validation in 3a2a87c

Reply:
Nice :)